crates/rethnet_evm/src/debug_trace.rs

```diff
@@ -0,0 +1,432 @@
+use crate::blockchain::SyncBlockchain;
+use crate::evm::build_evm;
+use crate::state::SyncState;
+use crate::{PendingTransaction, TransactionError};
+use rethnet_eth::signature::SignatureError;
+use rethnet_eth::B256;
+use revm::inspectors::GasInspector;
+use revm::interpreter::{
+    opcode, CallInputs, CreateInputs, Gas, InstructionResult, Interpreter, Stack,
+};
+use revm::primitives::{hex, B160, U256};
+use revm::primitives::{BlockEnv, Bytes, CfgEnv, ExecutionResult, ResultAndState, SpecId};
+use revm::{EVMData, Inspector, JournalEntry};
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+/// Get trace output for `debug_traceTransaction`
+#[cfg_attr(feature = "tracing", tracing::instrument)]
+pub fn debug_trace_transaction<BlockchainErrorT, StateErrorT>(
+    blockchain: &dyn SyncBlockchain<BlockchainErrorT, StateErrorT>,
+    // Take ownership of the state so that we can apply throw-away modifications on it
+    mut state: Box<dyn SyncState<StateErrorT>>,
+    evm_config: CfgEnv,
+    trace_config: DebugTraceConfig,
+    block_env: BlockEnv,
+    transactions: Vec<PendingTransaction>,
+    transaction_hash: &B256,
+) -> Result<DebugTraceResult, DebugTraceError<BlockchainErrorT, StateErrorT>>
+where
+    BlockchainErrorT: Debug + Send + 'static,
+    StateErrorT: Debug + Send + 'static,
+{
+    if evm_config.spec_id < SpecId::SPURIOUS_DRAGON {
+        // Matching Hardhat Network behaviour: https://github.com/NomicFoundation/hardhat/blob/af7e4ce6a18601ec9cd6d4aa335fa7e24450e638/packages/hardhat-core/src/internal/hardhat-network/provider/vm/ethereumjs.ts#L427
+        return Err(DebugTraceError::InvalidSpecId {
+            spec_id: evm_config.spec_id,
+        });
+    }
+
+    if evm_config.spec_id > SpecId::MERGE && block_env.prevrandao.is_none() {
+        return Err(TransactionError::MissingPrevrandao.into());
+    }
+
+    for transaction in transactions {
+        if transaction.hash() == transaction_hash {
+            let evm = build_evm(
```

Comment:
Could we build the `evm` outside of the look and just mutate the transaction?

Reply:
I tried, but haven't managed to figure this out in reasonable time. It's difficult, because of the need to `commit` the preceding transactions. In order to commit a transaction, we first need to give them `evm` a reference to the state:

```rust
evm.database(DatabaseComponents {
    state: &state,
    block_hash: blockchain,
});
```

And then commit the changes

```rust
// Needs `&mut state`
state.commit(changes)
```

And if the `evm` needs to live past performing a transaction, we won't be able to borrow the `state` as mutable to commit the result, because we gave the `evm` a shared reference to it in `evm.database(...)`.

Reply:
That makes sense. Thanks!